### DISCUSSION TITLE
Adding Attendant's Details

### DISCUSSION THREAD

**Discussion by oni-on1003** [opencrvs/opencrvs-core#7778] [Technical Q&A]
I'm not sure if this is the right section for this post.

I would like to capture the attendant's information, like name, occupation (title/designation), and address, and then display it on the certificate.

I tried to add a new section called "attendant" in country-config > src > form > birth > index.ts.

When I submit the form there's an error undefined "attendant" in type BirthRegistration

To fix the error I edit the following core file(sorry for this but I need to figure out how the system works)
- packages > gateway > src > features > registration > schema.graphql

`type BirthRegistration implements EventRegistration {
....,
attendant: Person # -> .section(code='attendant-details').entry.reference
}`

`
input BirthRegistrationInput {
....,
  attendant: PersonInput
}`
  
- packages > common > src > fhir > transformers > input.ts

`export interface BirthRegistration {
....,
attendant?: Person
}`

- packages > commons > src > fhir > transformers > index.ts

`import {
....,
ATTENDANT_CODE, 
ATTENDANT_TITLE
} from '..'`

`const builders: IFieldBuilders = {
....,
attendant: {
    _fhirID: (fhirBundle, fieldValue) => {
      const attendant = selectOrCreatePersonResource(
        ATTENDANT_CODE,
        ATTENDANT_TITLE,
        fhirBundle
      )
      attendant.id = fieldValue as string
    },
    identifier: createIDBuilder( ATTENDANT_CODE,  ATTENDANT_TITLE),
    name: createNameBuilder( ATTENDANT_CODE,  ATTENDANT_TITLE),
    address: createAddressBuilder( ATTENDANT_CODE,  ATTENDANT_TITLE),
  }
}`

- packages > commons > src > fhir > composition.ts

`export const ATTENDANT_TITLE = "Attendant's details"`
`export const ATTENDANT_CODE = 'attendant-details'`
`type ReferenceType =
.....
| {
      code: typeof ATTENDANT_CODE
      reference: Reference['reference']
      title: typeof ATTENDANT_TITLE
    }`

Unfortunately, I wasn't able to store the attendant's details. I hope someone can help me solve the problem. I'm currently using the v1.5 master.

If the OpenCRVS team finds it useful, I can make a pull request once I've finished it.

**Comment by tahmidrahman-dsi** [comment]
@oni-on1003 thanks for reaching out! 

Currently the form is consisted of a collection of specific fields. Some of which are 'required' and some are 'optional'. All are validated in the config form validation service [here](https://github.com/opencrvs/opencrvs-core/blob/6fd0d75a9416e85d6a58d41139371f528dcaccf7/packages/config/src/handlers/forms/validation.ts). 

Any addition to this set of fields, is implemented as 'custom' . In your case, you are going to add a 'custom' section, which means it should have all of its fields configured as custom. We can configure a field as custom by setting its `custom` property `true` value

**Reply by oni-on1003** [reply]
@tahmidrahman-dsi thank you for your response.

I'm able to make it work without setting the `custom` property of the fields to `true`.

I made some changes to other core files.

- packages > gateway > src > features > registration > type-resolvers.ts
  import the `ATTENDANT_CODE` and use it in the `findPatient` function then add this property `attendant: findPatient(ATTENDANT_CODE)` to `Birthregisration`
  
 - packages > client > src > views > ViewRecord > query.ts
  add the `attendant {...properties...}` to the query
  
- packages > client > src > views > DataProvider > birth > queries.ts
  add the `attendant {...properties...}` to the queries
  
 In summary, I've updated the following files:

- packages > gateway > src > features > registration > schema.graphql
- packages > gateway > src > features > registration > type-resolvers.ts
- packages > common > src > fhir > transformers > input.ts
- packages > commons > src > fhir > transformers > index.ts
- packages > commons > src > fhir > composition.ts
- packages > client > src > views > ViewRecord > query.ts
- packages > client > src > views > DataProvider > birth > queries.ts

Please let me know if the implementation is correct and if there's a file that I need to update.

Thank you

**Reply by tahmidrahman-dsi** [reply]
@oni-on1003 thanks for sharing your approach. 

Generally core modification is not recommended to configure forms, because it would be shared across different country implementations. For your case, adding "Attendant's Details" can be useful for your country's implementation but other countries might not need at all. In such case, the solution can be easily achieved by only configuring the country config form setting all the fields custom and it would work right away. 

We have opened a ticket so that the form validation prevents adding custom section without custom fields. 
https://github.com/opencrvs/opencrvs-core/issues/7788

I understand you have made an effort to make it work and it is great to learn that you closely followed the process with a deep insight, I would still suggest to configure only from the country config. Please let me know if you find any issues implementing this approach.

**Reply by oni-on1003** [reply]
@tahmidrahman-dsi I tried your suggestion but I'm getting an error when using the address decorator.
`BAD_USER_INPUT` `Variable \"$details\" got invalid value {....... Field \"attendant\" is not defined by type \"BirthRegistrationInput\"."`
Is there a way to set the address field as custom field for a custom section?

**Reply by tahmidrahman-dsi** [reply]
@oni-on1003 could you please try something like this in your form?
```
groups: [
  {
    id: 'attendant-details-view-group',
    fields: [
      createNameCustomField(),  // example custom field 1
      createBirthDateCustomField(),   // example custom field 2
      ...createAddressCustomFields()
    ],
    previewGroups: [attendantNameInEnglish]
  }
]
```
where `createAddressCustomFields` uses `getAddressFields` in the background turning the fields into custom ones
```
export function createAddressCustomFields(): SerializedFormField[] {
  return getAddressFields('attendantDetails', AddressCases.PRIMARY_ADDRESS).map(
    (field) => ({
      ...field,
      custom: true,
      customQuesstionMappingId: `birth.attendantDetails.attendant-details-view-group.${field.name}`,
      mapping: getCustomFieldMapping(
        `birth.attendantDetails.attendant-details-view-group.${field.name}`
      ),
      conditionals: [] // add any conditionals here
    })
  )
}
```

**Reply by oni-on1003** [reply]
@tahmidrahman-dsi thank you for the sample code.

I'm able to render the address fields but there's a problem when the form was submitted.

The payload `questionaire` doesn't have the `fieldId` of the address field
`[...... { "value": "c3021715-bb11-4285-bc7c-880535beb5ab" }, { "value": "7ba21f1b-4e90-437c-b401-e8cd3c1a5ec6" }, { "value": "some international state" }, { "value": "some international district" }, { "fieldId": "birth.mother.mother-view-group.motherIdType", "value": "NATIONAL_ID" }.... ]`

**Reply by oni-on1003** [reply]
@tahmidrahman-dsi  I found the culprit. there's a typo error `customQuesstionMappingId`.

Will update once I've finish my testing.

**Comment by oni-on1003** [comment]
@tahmidrahman-dsi Thank you for your help. It's working perfectly now.

Here's the function just in case someone needs it.
```
export function getCustomAddressFields(
  event: string,
  sectionId: string
): SerializedFormField[] {
  const addressFields = getAddressFields(
    sectionId,
    AddressCases.PRIMARY_ADDRESS
  ).map((field) => ({
    ...field,
    custom: true,
    customQuestionMappingId: `${event}.${sectionId}.${sectionId}-view-group.${field.name}`,
    mapping: getCustomFieldMapping(
      `${event}.${sectionId}.${sectionId}-view-group.${field.name}`
    ),
    conditionals: [
      {
        action: 'hide',
        expression: `(${field.name.indexOf('international')} !== -1)` //hide the international address fields, just remove this condition if not needed
      }
    ] // add any conditionals here
  }))
  return addressFields
}
``` 

Usage
```
groups: [
  {
    id: 'id-custom-of-view-group', //update the value of the id
    fields: [
      customField1(), //replace this with your custom fields
      ...getCustomAddressFields('event-name', 'section-name') // event-name = birth/death/marriage, section-name = your custom section name
   ],
  }
]
```
